### PR TITLE
Add Adaptive Personalized Federated Learning

### PR DESCRIPTION
**Pull request by Lando-L**
Adding the APFL algorithm from [Adaptive Personalized Federated Learning](https://doi.org/10.48550/arXiv.2003.13461).

### CODE REVIEW

File: experiments/apfl/run_apfl.py
Comment (jaehunro) [inline experiments/apfl/run_apfl.py:1]: Can you also add a README.md in `apfl/` with the summary of the experiment results so we can compare it to the paper? Also, if it's not too much work, I'm also interested in the speed/duration of the experiment and what devices/accelerators you're running on?

**Review by jaehunro — COMMENTED** [review]
Thanks for adding this algorithm! I left a couple of comments, but once those are addressed, it looks good to me!

File: experiments/apfl/run_apfl.py [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+
+  # Logging overall time it took.
+  num_rounds = config.num_rounds - start_round_num + 1
+  mean_round_duration = ((time.time() - start) / num_rounds if num_rounds > 0 else 0)
+
+  # Final evaluation.
+  final_eval_start = time.time()
+  metrics = eval_algorithm.apply(state, list(test_fd.clients()))
+  if metrics:
+    metrics_path = os.path.join(config.root_dir, 'full_test_eval.tsv')
```
[the remark this answers is not in the document]
Comment (Lando-L) [reply experiments/apfl/run_apfl.py]: I updated the evaluation function in the algorithm to return the `for_each_client` generator. This way in the experiment we are more flexible to aggregate the client metrics or log the metrics for each client individualy.

File: experiments/apfl/run_apfl.py
Comment (jaehunro) [reply experiments/apfl/run_apfl.py:1]: Fair enough. Just add a disclaimer at the end of your README just detailing what you wrote here.

File: fedjax/algorithms/apfl.py [hunk 2]
```diff
@@ -239,83 +237,69 @@ def create_eval_for_each_client(model: models.Model):
...
+  [ServerState, Sequence[Tuple[federated_data.ClientId, client_datasets.ClientDataset]]],
+  Sequence[Tuple[federated_data.ClientId, Mapping[str, jnp.array]]]
+]:
   """Evaluates adaptive personalized federated learning.
 
   Args:
     model: Model
-    client_batch_hparams: Hyperparameters for batching client dataset for train.
+    client_batch_hparams: Hyperparameters for batching client dataset for eval.
+    aggregate: Flag whether eval metrics should be aggregated over clients.
```
Comment (jaehunro) [inline fedjax/algorithms/apfl.py:281]: This arg description for `aggregate` looks like it was left in by accident since I don't see it in the args list.

**Review by jaehunro — COMMENTED** [review]
Just a couple of small comments. Other than that, it looks good to me!
Thanks!

**Review by jaehunro — COMMENTED** [review]
Thanks! One last comment (I should've realized earlier): it seems that some of your py files have lines that are just white spaces. Can you clean this up with `sed -i -e 's/^ *$//'` ?

Thanks!

**Comment by Lando-L** [comment]
@jaehunro Done! Thank you for the feedback :)

**Pull request merged**